### ISSUE TITLE
trait type to distinguish generic offset array and one based indexing array

### ISSUE THREAD

**Issue by johnnychen94** [JuliaArrays/ArrayInterface.jl#326]
Traits like this might be helpful to be kept in ArrayInterfaceCore

```julia
struct GenericIndexing end
struct OneBasedIndexing end

offset_type(::Array) = OneBasedIndexing()
offset_type(::OffsetArray) = GenericIndexing()
offset_type(A::Diagonal) = offset_type(parent(A))
```

CRef: https://discourse.julialang.org/t/does-abstractarray-itself-need-to-support-indexing-beyond-1-based/83324/3

**Comment by Tokazama** [comment]
If you need something right away you could probably accomplish this by doing `is_offset_from_parent(T::Type) = any(!=(1), known_offsets(parent_type(T)))`, since the only time `known_offsets` doesn't return ones is when dealing with non one-based indexing.

In general, it would be helpful to have more tools that assist in transforming between index states so that when wrapper like `Diagonal` defines its own indexing rules we know what is safe/optimal to pass off to the parent array.

**Comment by johnnychen94** [comment]
I started to be skeptical about this proposal now as it only provides a static alternative to the current `Base.require_one_based_indexing`. And the benefit seems doesn't worth the efforts. -- I somehow believe Julia compiler could just optimize this runtime axes check away.

I've outlined the reasoning in the discourse discussion https://discourse.julialang.org/t/does-abstractarray-itself-need-to-support-indexing-beyond-1-based/83324/6?u=johnnychen94
